### PR TITLE
Implement CubeHelix sampling function

### LINKED ISSUE #30: add 'cubehelix' colour scheme
I came across ['cubehelix' colour scheme](https://www.mrao.cam.ac.uk/~dag/CUBEHELIX/).
[This page](https://www.mrao.cam.ac.uk/~dag/CUBEHELIX/cubetry.html) shows how different parameters effect the scheme. it would be great if we can reconstract that configuration in some object and then define something like `cubehelix:: CubehelixConfig -> ColorScale`.
[Here is python](https://github.com/jradavenport/cubehelix/blob/master/cubehelix.py) implementation, on original page there are cople more.

### PR DESCRIPTION
**Pull request by safareli**
fix #30

### CODE REVIEW

File: src/Color.purs [hunk 1]
```diff
@@ -538,6 +546,32 @@ mix Lab c1 c2 frac = lab
     f = toLab c1
     t = toLab c2
 
+radians :: Radians
+radians = pi / 180.0
+
+cubehelixMix :: Number -> Mixer -- TODO fix alpha
+cubehelixMix gama (HSLA ah' as' al' _) (HSLA bh' bs' bl' _) =
```
Comment (safareli) [inline src/Color.purs]: ported from https://github.com/d3/d3-plugins/blob/master/cubehelix/cubehelix.js#L13

File: src/Color.purs [hunk 2]
```diff
@@ -538,6 +546,32 @@ mix Lab c1 c2 frac = lab
...
+  let
+    ah = (ah' + 120.0) * radians
+    bh = (bh' + 120.0) * radians - ah
+    as = as'
+    bs = bs' - as
+    al = al'
+    bl = bl' - al
+
+  -- NOTE not sure why isNan check is needed so I have not ported it
```
Comment (safareli) [inline src/Color.purs]: I will remove this comments, as still not sure why it's needed 🗡

File: src/Color.purs [hunk 3]
```diff
@@ -538,6 +546,32 @@ mix Lab c1 c2 frac = lab
...
+  -- if (isNaN(bs)) bs = 0, as = isNaN(as) ? b.s : as;
+  -- if (isNaN(bh)) bh = 0, ah = isNaN(ah) ? b.h : ah;
+  in \t ->
```
Comment (safareli) [inline src/Color.purs]: I think it should be much better in performance, what's your thoughts?

File: src/Color/Scale.purs [hunk 4]
```diff
@@ -53,19 +64,26 @@ stopColor :: ColorStop -> Color
 stopColor (ColorStop c _) = c
 
 -- | A color scale.
-data ColorScale = ColorScale ColorSpace Color (List ColorStop) Color
+data ColorScaleBuilder = ColorScaleBuilder Color (List ColorStop) Color
```
Comment (safareli) [inline src/Color/Scale.purs]: Exporting this is very useful so users could write their own helper functions, without waiting for their PR to be merged. Before I needed to define my own type for it [see this](https://github.com/safareli/purescript-colorpalettepicker-halogen/blob/56a98be084c35e596438187aa1fc0152e5d8bb9a/src/ColorPalettePicker/Utils/PreScale.purs#L10)

Also if we have separate type which does not contain `ColorSpace` then we can implement `combineScale :: Number → Scale → Scale → Scale` which i did in the link above. If the Scale contained colorSpace then we would have to discard one of them, which I think will not be a good thing to do.

File: src/Color/Scale.purs [hunk 5]
```diff
@@ -53,19 +64,26 @@ stopColor :: ColorStop -> Color
...
+colorScale' :: Color -> List ColorStop -> Color -> ColorScaleBuilder
+colorScale' = ColorScaleBuilder
 
 -- | Create a uniform color scale from a list of colors that will be evenly
 -- | spaced on the scale.
 uniformScale :: forall f. Foldable f => ColorSpace -> Color -> f Color -> Color -> ColorScale
-uniformScale mode b middle e = colorScale mode b stops e
+uniformScale mode b middle e = ColorScale mode $ uniformScale' b middle e
+
+uniformScale' :: forall f. Foldable f => Color -> f Color -> Color -> ColorScaleBuilder
```
Comment (safareli) [inline src/Color/Scale.purs]: this `'` functions might be useful when someone is working on ColorScaleBuilder and might not care about ColorSpace. Like I did [here](https://github.com/safareli/purescript-colorpalettepicker-halogen/blob/9b47d80d1bec6a5e44eb1361045b24166d0d0e35/src/ColorPalettePicker/Palettes.purs#L294)

**Comment by safareli** [comment]
Stack overflows on `HSV -> HSL -> HSV, HSL -> HSV -> HSL` test :/

**Comment by sharkdp** [comment]
Thank you for working on this. Let me know when it's ready to review.

**Comment by safareli** [comment]
It's ready for review

File: src/Color/Scale.purs [hunk 6]
```diff
@@ -21,18 +21,29 @@ module Color.Scale
   , blueToRed
   , yellowToRed
   , cssColorStops
+
+  -- more advanced staff
```
Comment (sharkdp) [inline src/Color/Scale.purs]: Can we choose a more useful comment here? Also, do we have to export all of these functions?

File: src/Color/Scale.purs (hunk 4 above)
Comment (sharkdp) [reply src/Color/Scale.purs]: :+1:

Comment (sharkdp) [reply src/Color/Scale.purs]: What about just naming this `ColorStops` or `ColorStopList`?

File: src/Color/Scale.purs [hunk 7]
```diff
@@ -53,19 +64,26 @@ stopColor :: ColorStop -> Color
...
 
 -- | Create a color scale. The color space is used for interpolation between
 -- | different stops. The first `Color` defines the left end (color at ratio
 -- | 0.0), the list of stops defines possible intermediate steps and the second
 -- | `Color` argument defines the right end point (color at ratio 1.0).
 colorScale :: ColorSpace -> Color -> List ColorStop -> Color -> ColorScale
-colorScale = ColorScale
+colorScale space b middle e = ColorScale space $ colorScale' b middle e
+
+colorScale' :: Color -> List ColorStop -> Color -> ColorScaleBuilder
```
Comment (sharkdp) [inline src/Color/Scale.purs]: This, in turn, could then be named `colorStops` or `colorStopList`.

File: src/Color/Scale.purs (hunk 5 above)
Comment (sharkdp) [reply src/Color/Scale.purs]: Okay. Can we please add a docstring?

File: src/Color/Scale.purs [hunk 8]
```diff
@@ -75,15 +93,29 @@ uniformScale mode b middle e = colorScale mode b stops e
...
+
+addStop' :: ColorScaleBuilder -> Color -> Number -> ColorScaleBuilder
+addStop' (ColorScaleBuilder b middle e) c r =
+  ColorScaleBuilder b (insertBy (comparing stopRatio) stop middle) e
     where stop = colorStop c r
 
+sample :: Sampler ColorScale
+sample (ColorScale mode scale) = mkSimpleSampler (mix mode) scale
+
+type Sampler a = a -> Number -> Color
```
Comment (sharkdp) [inline src/Color/Scale.purs]: IMHO: this type synonym just adds mental overhead when reading the type definitions below.

File: src/Color/Scale.purs [hunk 9]
```diff
@@ -75,15 +93,29 @@ uniformScale mode b middle e = colorScale mode b stops e
...
+
+
+
+mkSimpleSampler :: Mixer -> Sampler ColorScaleBuilder
 -- | Get the color at a specific point on the color scale (number between 0 and
 -- | 1). If the number is smaller than 0, the color at 0 is returned. If the
 -- | number is larger than 1, the color at 1 is returned.
-sample :: ColorScale -> Number -> Color
-sample scale@(ColorScale mode b middle e) x
+mkSimpleSampler mixer (ColorScaleBuilder b middle e) x
```
Comment (sharkdp) [inline src/Color/Scale.purs]: Why is there a docstring between type declaration and function implementation?

File: src/Color/Scale.purs [hunk 10]
```diff
@@ -93,23 +125,30 @@ sample scale@(ColorScale mode b middle e) x
...
 colors :: ColorScale -> Int -> List Color
-colors (ColorScale _ _ _ _) 0 = Nil
-colors (ColorScale _ b _ _) 1 = singleton b
-colors scale num = do
-  i <- 0 .. (num - 1)
-  pure $ sample scale (toNumber i / toNumber (num - 1))
+colors scale = colors' (sample scale)
+
+-- | Takes sampling function and returns A list of colors that is sampled from
+-- | a color scale. The number of colors can be specified.
```
Comment (sharkdp) [inline src/Color/Scale.purs]: Does `colors'` have to be a top-level function?

File: src/Color/Scale.purs [hunk 11]
```diff
@@ -176,24 +215,28 @@ cool = colorScale RGB (hsl 180.0 1.0 0.6) Nil (hsl 300.0 1.0 0.5)
...
-    <> intercalate ", " (toString <$> middle)
-    <> ", " <> cssStringHSLA e
-  where
-    toString (ColorStop c r) = cssStringHSLA c <> " " <> percentage r
-    percentage r = show (r * 100.0) <> "%"
+cssColorStops (ColorScale RGB b) = cssColorStopsRGB b
+cssColorStops (ColorScale _ scale) = cssColorStopsSample (mkSimpleSampler $ mix RGB) 10 scale
 
-cssColorStops scale@(ColorScale _ b middle e) = cssColorStops csRGB
+cssColorStopsSample :: Sampler ColorScaleBuilder -> Int -> ColorScaleBuilder -> String
```
Comment (sharkdp) [inline src/Color/Scale.purs]: What are these for? Can you please add some comments/documentation?

File: src/Color.purs [hunk 12]
```diff
@@ -167,12 +170,14 @@ rgb' r g b = rgba' r g b 1.0
 -- | Hue is given in degrees, as a `Number` between 0.0 and 360.0. Saturation,
 -- | Lightness and Alpha are numbers between 0.0 and 1.0.
 hsla :: Number -> Number -> Number -> Number -> Color
-hsla h s l a = HSLA h' s' l' a'
-  where h' = if h == 360.0 then h else h `modPos` 360.0
-        s' = clamp 0.0 1.0 s
+hsla h s l a = HSLA h s' l' a'
+  where s' = clamp 0.0 1.0 s
```
[the remark this answers is not in the document]
Comment (safareli) [reply src/Color.purs]: it's needed for coubehelix mix of colors.
for example if we want gradient from volor eith hue 350 to color with hue 370.
without this change we will get gradient from 350 down to 10 where wi will have length of 340 in hue, not up to 370 where length will be 20. and the distinction between 370 and 10 might be useful for other sampling functions too

to not forget cliping we could use toHSLA and do not patern match on constructor, or sormthing else.

[the remark this answers is not in the document]
Comment (safareli) [reply src/Color.purs]: you can revert this part and see that coubhelix gradient will look like

[the remark this answers is not in the document]
Comment (sharkdp) [reply src/Color.purs]: I see, okay.

[the remark this answers is not in the document]
Comment (safareli) [reply src/Color.purs]: I have just pushed a refacto of this issue.
let me know if you like the approach, or if you can think of different approach to issue?

[the remark this answers is not in the document]
Comment (sharkdp) [reply src/Color.purs]: In this case, I rather prefer the previous approach, to be honest. With both `hue` and `hueOriginal` as fields, we have to keep track of always keeping both in sync. Sorry :smile: 

Thinking about this a bit more, I guess it's okay to allow for angles outside of `[0°, 360°]`.

[the remark this answers is not in the document]
Comment (safareli) [reply src/Color.purs]: I've reverted most of the changes but made this change, so we are not forced to remember clipping Hue:
```purescript
data Color = HSLA Hue Number Number Number

newtype Hue = UnclippedHue Number

clipHue :: Hue -> Number
clipHue (UnclippedHue x) = x `modPos` 360.0
```

[the remark this answers is not in the document]
Comment (sharkdp) [reply src/Color.purs]: :+1:

File: src/Color.purs [hunk 13]
```diff
@@ -500,12 +508,17 @@ interpolateAngle fraction a b = interpolate fraction shortest.from shortest.to
...
+type Interpolator = Color -> Color -> Number -> Color
+
 -- | Mix two colors by linearly interpolating between them in the specified
 -- | color space. For the HSL colorspace, the shortest path is chosen along the
 -- | circle of hue values.
-mix :: ColorSpace -> Color -> Color -> Number -> Color
+mix :: ColorSpace -> Interpolator
 mix HSL c1 c2 frac = hsla
-    (interpolateAngle frac f.h t.h)
+    (interpolateAngle frac f.h t.h) -- NOTE here we might benefit from using uncliped Hue?
```
Comment (safareli) [inline src/Color.purs]: Do you see use of UnclipedHue here?

File: src/Color.purs [hunk 14]
```diff
@@ -538,6 +551,33 @@ mix Lab c1 c2 frac = lab
     f = toLab c1
     t = toLab c2
 
+
+-- | Mix two colors by [Dave Green's `cubehelix'](http://www.mrao.cam.ac.uk/~dag/CUBEHELIX/) interpolation between them.
+-- | Takes gamma correction value as an argument and retunrs Interpolator.
+-- | For more details see: [d3-plugins/cubehelix](https://github.com/d3/d3-plugins/tree/40f8b3b91e67719f58408732d7ddae94cafa559a/cubehelix#interpolateCubehelix)
```
Comment (safareli) [inline src/Color.purs]: do links like this work in pursuit?

Comment (sharkdp) [reply src/Color.purs]: I think so.

**Comment by safareli** [comment]
I have added comments to new functions, let me know if you spot some typos or would like different wording.

File: src/Color.purs (hunk 13 above)
Comment (sharkdp) [reply src/Color.purs]: Yes, good idea!

**Comment by sharkdp** [comment]
I don't have any time today but I'll try to review it in full tomorrow. Thank you very much.

Do you know why the tests started failing now? Why did the stackoverflow only appear now?

**Comment by safareli** [comment]
> Do you know why the tests started failing now? Why did the stackoverflow only appear now?

Not sure, when I was adding the test which is now failing it was raising stackoverflow at first  and because of that I used smaller numbers.

**Comment by safareli** [comment]
I just found out that it's allowed to have hue be out of range [0,360]

https://www.w3.org/TR/css3-color/#hsl-color
> HSL colors are encoding as a triple (hue, saturation, lightness). Hue is represented as an angle of the color circle (i.e. the rainbow represented in a circle). This angle is so typically measured in degrees that the unit is implicit in CSS; syntactically, only a <number> is given. By definition red=0=360, and the other colors are spread around the circle, so green=120, blue=240, etc. As an angle, it implicitly wraps around such that -120=240 and 480=120. One way an implementation could normalize such an angle x to the range [0,360) (i.e. zero degrees, inclusive, to 360 degrees, exclusive) is to compute (((x mod 360) + 360) mod 360). Saturation and lightness are represented as percentages. 100% is full saturation, and 0% is a shade of gray. 0% lightness is black, 100% lightness is white, and 50% lightness is “normal”.

This makes me think that we should  remove clipping from `cssStringHSLA`. what you think?


Besides `cssStringHSLA` we call clip in `to{RGB,HSL,HSV}A`. 
Also There is no way for user to get unclipped Hue.  I think we should add  `toHSLA'` or similar, which returns `{ h :: Hue, s :: Number, l :: Number, a :: Number }` and export `Hue` + `clipHue`. thoughts?

**Comment by safareli** [comment]
I came to `d3/d3-color` here cubehelix is implemented [as color space](https://github.com/d3/d3-color#cubehelix), not as just an interpolation. 

If we try same thing I see it as creating `Color.cobehelix` which takes `h s l a` which converts CobeHelix color to HSLA to construct Color object. Then adding `CobeHelix` to `Color.ColorSpace` and we will add case for this colorspace in `Color.mix`, where Colors `a` `b` will be converted to CobeHelix mixed and then back to HSLA. 

But I don't quite see how to implement, will need to dig into [implementation](https://github.com/d3/d3-color/blob/8880bbce23bfdb05d4774a4d459c15d64f7e56ea/src/cubehelix.js)
Will I do not get is what's the range of saturation, so I think I will stop on this until i hear back on https://github.com/d3/d3-color/issues/34

---

[here](https://gist.github.com/mbostock/1009e8fa6beae97d6f19) is a:
> A two-dimensional cross section of the Cubehelix color space: from top to bottom, hue ranges from 0° to 360°; from left to right, lightness ranges from 1 to 0. Saturation is a constant 1.
Raw

if you try to look at that image with grayscale enabled with on our PC/mobile you will see that lightness is monotonic.

[This](https://bl.ocks.org/mbostock/ba8d75e45794c27168b5) is Color picker in CobeHelix space.

**Comment by sharkdp** [comment]
> Not sure, when I was adding the test which is now failing it was raising stackoverflow at first and because of that I used smaller numbers.

Thanks for fixing those

> This makes me think that we should remove clipping from cssStringHSLA. what you think?

Yeah, good point. Thanks for the reference.

> Besides cssStringHSLA we call clip in to{RGB,HSL,HSV}A.
> Also There is no way for user to get unclipped Hue. I think we should add toHSLA' or similar, which returns { h :: Hue, s :: Number, l :: Number, a :: Number } and export Hue + clipHue. thoughts?

Maybe it's actually best if we return unclipped values everywhere. Do we need the clipped values at some point? :smile:



> But I don't quite see how to implement, will need to dig into implementation
> Will I do not get is what's the range of saturation, so I think I will stop on this until i hear back on d3/d3-color#34

Cool. Thanks for looking into this. Sounds like a good option to me (adding it as a ColorSpace).

**Comment by safareli** [comment]
> Do we need the clipped values at some point?

We need to clip for `toRGBA`, on the other side for some applications clipped hue might be needed. So what you are suggesting is to just store unclipped hue as Number and expose `clipHue :: Number -> Number` which will be used only in `toRGBA`and in client code ? or what

About `ColorSpace` I don't yet understand how we can do that I have some open questions in the issue will take a look into it in future maybe. 

So to merge this we just need to decide what should we don on clipping.

**Comment by sharkdp** [comment]
> We need to clip for toRGBA, on the other side for some applications clipped hue might be needed. So what you are suggesting is to just store unclipped hue as Number and expose clipHue :: Number -> Number which will be used only in toRGBAand in client code ? or what

I think it's perfectly fine the way you did it right now, thank you very much.

**Comment by sharkdp** [comment]
@safareli 

I have added `cubehelix :: ColorScale`. Thought you might find this useful.

I've also added the `mixCubehelix` function and the `cubehelix` colorscale to the interactive documentation: http://sharkdp.github.io/purescript-colors/

**Comment by safareli** [comment]
nice!

**Comment by safareli** [comment]
Let me know when you release new version :P

**Comment by sharkdp** [comment]
> Let me know when you release new version :P

Done. Thanks again!

**Comment by sharkdp** [comment]
I've also added a small interactive cubehelix experiment at the bottom of the page:
http://sharkdp.github.io/purescript-colors/

(similar to https://www.mrao.cam.ac.uk/~dag/CUBEHELIX/cubetry.html)

**Comment by safareli** [comment]
sweet 🍓

**Pull request merged**